bot.py

```diff
@@ -82,47 +136,92 @@ def start_game_msg(name):
-        time.sleep(1)
-    click_image("rc_items/gain_power.png")
+def end_game(self, fail=False):
+    if not fail:
+        self.game_status = "idle"
+
+        keyboard.press_and_release("page up")
+        keyboard.press_and_release("down")
+
+        while not check_image("rc_items/gain_power.png"):
+            time.sleep(1)
+        click_image("rc_items/gain_power.png")
+
+        time.sleep(3)
+
+        keyboard.press_and_release("page up")
+        time.sleep(2)
+        click_image("rc_items/goto_games.png")
+        time.sleep(2)
 
-    keyboard.press_and_release("page up")
-    time.sleep(0.5)
-    click_image("rc_items/goto_games.png")
-    time.sleep(0.5)
-    if check_image("rc_items/collect_pc.png"):
-        click_image("rc_items/click_image")
+        if check_image("rc_items/collect_pc.png"):
+            click_image("rc_items/click_image")
+    else:
+        keyboard.press_and_release("page up")
+        time.sleep(2)
+
+        click_image("rc_items/goto_games.png")
+
+        os.execv(sys.executable, ['python'] + sys.argv) #restart script
+
+
+class ThreadWithReturnValue(Thread):
+    def __init__(self, group=None, target=None, name=None,
+                 args=(), kwargs={}, Verbose=None):
+        Thread.__init__(self, group, target, name, args, kwargs)
+        self._return = None
+
+    def run(self):
+        if self._target is not None:
+            self._return = self._target(*self._args,
+                                        **self._kwargs)
+
+    def join(self, *args):
+        Thread.join(self, *args)
+        return self._return
 
 
 class Bot2048:
     def __init__(self):
         self.start_img_path = "rc_items/2048_gameimg.png"
         self.available_moves = ["right", "left", "up", "down"]
         self.game = "2048"
+        self.game_status = "idle"
 
     def can_start(self):
         return check_image(self.start_img_path)
 
     def play(self):
-        err = start_game(self.start_img_path)
+        err = start_game(self, self.start_img_path)
         if err:
             return not err
         start_game_msg(self.game)
         self.run_game()
-        end_game()
+        end_game(self)
 
     def run_game(self):
-        while not check_image("rc_items/gain_power.png"):
-            for _ in range(4):
+        self.game_status = "running"
+
+        try:
+            thread = ThreadWithReturnValue(target=check_image,
+                                           args=("rc_items/gain_power.png", True, self,))
+            thread.start()
+        except:
+            print("Unable to start thread for checking image")
+            end_game(self, fail=True)
+
+        while self.game_status == "running":
+            for i in range(8):
                 keyboard.press_and_release(random.choice(self.available_moves))
-                time.sleep(0.125)
+                time.sleep(0.15)
+            keyboard.press_and_release("page up")  # to prevent errors for the thread with check image
```

Comment:
checks if game status is still "running" (is set to "ended" if thread found the "gain_power" button)
and then pressing random button 8 times before checking status again to improve the speed